File: rust/src/day06.rs
use std::ops::Div;

use crate::{
    check_result,
    utils::{Context, Part},
};

struct Race {
    time: f64,
    distance: f64,
}
fn parse(lines: &Vec<String>, is_part_two: bool) -> Vec<Race> {
    let times_str = lines[0].split_once(":").unwrap().1.split_ascii_whitespace();
    let distances_str = lines[1].split_once(":").unwrap().1.split_ascii_whitespace();

    if is_part_two {
        let time = times_str
            .collect::<Vec<&str>>()
            .join("")
            .parse::<f64>()
            .unwrap();
        let distance = distances_str
            .collect::<Vec<&str>>()
            .join("")
            .parse::<f64>()
            .unwrap();
        return vec![Race { time, distance }];
    }

    let times = times_str.map(|n| n.parse::<f64>().unwrap());
    let distances = distances_str.map(|n| n.parse::<f64>().unwrap());

    return times
        .zip(distances)
        .map(|(time, distance)| Race { time, distance })
        .collect();
}

fn calculate_roots(race: &Race)-> (f64,f64) {
    let delta = (race.time.powi(2) - race.distance*4.0).sqrt();
    let min = (race.time - delta).div(2.0); 
    let max = (race.time + delta).div(2.0);
    return (min, max );
}

const PRECISION:f64 = 0.000001;
fn solve(race: &Race)-> u32 {
    let (x1,x2) = calculate_roots(race);
    let (x1_rounded,x2_rounded) = (x1.ceil(), x2.floor());
    let min_time:u32 = (x1_rounded + if (x1_rounded-x1).abs()<PRECISION {1.0} else { 0.0 }) as u32;
    let max_time:u32 = (race.time.min(x2_rounded - if (x2_rounded-x2).abs() <PRECISION { 1.0} else { 0.0 })) as u32;
    let solution = max_time - min_time + 1;
    return solution;
}


pub fn puzzle(context: &Context, lines: &Vec<String>) {
    let values = parse(lines, context.is_part(Part::Part2));

    let result = values.iter().map(solve).fold(1,|a, b| a * b);
    
    if context.is_part(Part::Part1) {
        check_result!(context, result, [288, 4403592]);
    } else {
        check_result!(context, result, [71503, 38017587]);
    }
}

File: history/day_06.ts
import { Logger, Part, run, Type } from "../day_utils"

function parseLine(line: string, part: Part): number[] {
    const numbers = line.split(/:\s+/)[1];
    if (part === Part.PART_1) {
        return numbers.split(/\s+/).map(part => parseInt(part, 10));
    }
    return [parseInt(numbers.replaceAll(/\s+/g, ""), 10)];
}
interface Race { time: number, distance: number };


function parse(lines: string[], part: Part): Race[] {
    const times = parseLine(lines[0], part)
    const distances = parseLine(lines[1], part);
    return times.map((time, i) => {
        return {
            time,
            distance: distances[i]
        };
    });
}

function calculateRoots(race: Race): [number, number] {
    const delta = Math.sqrt(race.time ** 2 - 4 * race.distance);
    return [(race.time - delta) / 2, (race.time + delta) / 2 - Number.MIN_VALUE];
}

function solve(race: Race): number {
    const roots = calculateRoots(race);
    const roundedRoots = [Math.ceil(roots[0]), Math.floor(roots[1])];
    const maxTime = Math.min(race.time, (roundedRoots[1] === roots[1]) ? roundedRoots[1] - 1 : roundedRoots[1]);
    const minTime = (roundedRoots[0] === roots[0]) ? roundedRoots[0] + 1 : roundedRoots[0];
    const solution = maxTime - minTime + 1;
    return solution;
}

function puzzle(lines: string[], part: Part, type: Type, logger: Logger): void {
    const data = parse(lines, part);
    const result = data.map(race => solve(race)).reduce((a, b) => a * b, 1);;
        
    if (part === Part.PART_1) {
        logger.result(result, [288, 4403592])
    }
    else {
        logger.result(result, [71503, 38017587])
    }
}

/**
 * Update the date number after copy
 * Adapt types list to your needs and parts also 
 */
run(6, [Type.TEST, Type.RUN], puzzle, [Part.PART_1, Part.PART_2])